orchestrator/careplanservice/policy/agent.go

```diff
@@ -0,0 +1,199 @@
+package policy
+
+import (
+	"context"
+	_ "embed"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	fhirclient "github.com/SanteonNL/go-fhir-client"
+	"github.com/SanteonNL/orca/orchestrator/lib/auth"
+	"github.com/SanteonNL/orca/orchestrator/lib/coolfhir"
+	"github.com/open-policy-agent/opa/v1/rego"
+	"github.com/rs/zerolog"
+	"github.com/zorgbijjou/golang-fhir-models/fhir-models/fhir"
+)
+
+func subjectToParams(subject *fhir.Reference) (url.Values, error) {
+	params := url.Values{}
+
+	if coolfhir.IsLogicalReference(subject) {
+		params["patient:Patient.identifier"] = []string{fmt.Sprintf("%s|%s", *subject.Identifier.System, *subject.Identifier.Value)}
+	} else if subject.Id != nil {
+		params["subject"] = []string{fmt.Sprintf("Patient/%s", *subject.Id)}
+	} else if subject.Reference != nil {
+		params["subject"] = []string{*subject.Reference}
+	} else {
+		return params, fmt.Errorf("invalid subject (subject=%+v)", subject)
+	}
+
+	return params, nil
+}
+
+type SearchCache map[*fhir.Reference][]fhir.CarePlan
+
+func NewSearchCache() SearchCache {
+	return SearchCache{}
+}
+
+//go:embed policy.rego
+var source string
+
+var BuiltinModule = RegoModule{
+	Package: "policy",
+	Source:  source,
+}
+
+var ErrAccessDenied = errors.New("request denied by policy")
+
+type Agent struct {
+	logger zerolog.Logger
+	query  rego.PreparedEvalQuery
+	client fhirclient.Client
+}
+
+type RegoModule struct {
+	Package string
+	Source  string
+}
+
+type Context struct {
+	Principal fhir.Identifier `json:"principal"`
+	Method    string          `json:"method"`
+	Roles     []string        `json:"roles"`
+	Resource  any             `json:"resource"`
+	CarePlans []fhir.CarePlan `json:"careplans"`
+}
+
+func NewAgent(ctx context.Context, module RegoModule, client fhirclient.Client) (Agent, error) {
+	r := rego.New(
+		rego.Query(fmt.Sprintf("allow = data.%s.allow", module.Package)),
+		rego.Module(module.Package, module.Source),
+	)
+	query, err := r.PrepareForEval(ctx)
+	if err != nil {
+		return Agent{}, err
+	}
+
+	return Agent{
+		query:  query,
+		client: client,
+		logger: zerolog.New(os.Stdout),
```

Comment:
can you derive the logger from `ctx`? Then contextual log fields are added as well.